collect_app/src/main/java/org/odk/collect/android/widgets/AbstractSelectOneWidget.java

```diff
@@ -85,148 +76,97 @@ public AbstractSelectOneWidget(Context context, FormEntryPrompt prompt, boolean
-            button.cancelLongPress();
-        }
-    }
-
-    protected RadioButton createRadioButton(int index) {
-        String choiceName = getFormEntryPrompt().getSelectChoiceText(items.get(index));
-        CharSequence choiceDisplayName;
-        if (choiceName != null) {
-            choiceDisplayName = TextUtils.textToHtml(choiceName);
-        } else {
-            choiceDisplayName = "";
-        }
-
-        AppCompatRadioButton radioButton = new AppCompatRadioButton(getContext());
-        radioButton.setTextSize(TypedValue.COMPLEX_UNIT_DIP, getAnswerFontSize());
-        radioButton.setText(choiceDisplayName);
-        radioButton.setMovementMethod(LinkMovementMethod.getInstance());
-        radioButton.setTag(index);
-        radioButton.setId(ViewIds.generateViewId());
-        radioButton.setEnabled(!getFormEntryPrompt().isReadOnly());
-        radioButton.setFocusable(!getFormEntryPrompt().isReadOnly());
-        radioButton.setOnClickListener(this);
-
-        //adapt radioButton text as per language direction
-        if (isRTL()) {
-            radioButton.setGravity(Gravity.END);
-        } else {
-            radioButton.setGravity(Gravity.START);
-        }
-
-        if (items.get(index).getValue().equals(selectedValue)) {
-            radioButton.setChecked(true);
-        }
-
-        radioButton.setOnCheckedChangeListener(this);
+        SelectChoice selectChoice =  adapter.getSelectedItem();
 
-        return radioButton;
+        return selectChoice == null
+                ? null
+                : this instanceof ItemsetWidget
+                    ? new StringData(selectChoice.getValue())
+                    : new SelectOneData(new Selection(selectChoice));
     }
 
     protected void createLayout() {
         readItems();
 
-        LayoutInflater inflater = LayoutInflater.from(getContext());
+        adapter = new SelectOneListAdapter(items, selectedValue, this);
 
         if (items != null) {
-            for (int i = 0; i < items.size(); i++) {
-                answerLayout.addView(createRadioButtonLayout(inflater, i));
+            RecyclerView recyclerView = new RecyclerView(getContext());
+            recyclerView.setHasFixedSize(true);
+            recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
+            recyclerView.addItemDecoration(new DividerItemDecoration(getContext(), DividerItemDecoration.VERTICAL));
+            recyclerView.setAdapter(adapter);
+            answerLayout.addView(recyclerView);
+
+            // Only let the RecyclerView take up 100% of the screen height
+            DisplayMetrics displayMetrics = new DisplayMetrics();
+            try {
+                ((FormEntryActivity) getContext()).getWindowManager().getDefaultDisplay().getMetrics(displayMetrics);
+            } catch (ClassCastException e) {
+                Timber.i(e);
             }
+            if (recyclerView.getLayoutParams().height > displayMetrics.heightPixels) {
```

Comment:
You are right sorry for that mistake.
I think calculating a single list element is not possible because we may have items with different heights (you know media files, photos, different heights). So we wouldn't be able to perform a reliable calculation. Nothing else comes to my mind... I added a solution with 30 items as a cutoff value. I guess there is no screen that would be able to display 30 or more items without scroll.